### ISSUE TITLE
SOC is not working correctly

### ISSUE THREAD

**Issue by Ederborg** [OpenSource-EBike-firmware/Color_LCD#18]
If i set Battery total to 1000 and Used to 70 i get 70% used and 30% left. I think the battery total or the percentage is off by 10 times the value. Maybe it's a simple bug to fix?

I will set mine to 8500 (850 wh in reality) and try that for a while.

This is on my sw102

**Comment by geeksville** [comment]
Alas - I don't know much about the SOC calculations, but this code is common between the sw102 and 850C.  I'm going to be away for a few weeks.  @casainho can you look at it when you get a chance?

**Comment by lowPerformer** [comment]
Can confirm this bug even on rather old version 19.11.
LCD3 works, maybe we can compare those two code bases. Where is the SOC magic happening?

**Comment by casainho** [comment]
battery_soc() prints on main screen the SOC, print the variable ui16_m_battery_soc_watts_hour.

l2_calc_wh() calcs the value and is time dependent, it assumes it is called every 100ms. It is called then from layer_2(), that in 850C is called from TIM4_IRQHandler() but on SW102 is called from gui_timer_timeout(). Looking at the code, seems that gui_timer_timeout() is called every 20ms and inside is this code which makes layer_2() be called every 100ms as expected:

```
  gui_ticks++;

  if(gui_ticks % (100 / MSEC_PER_TICK) == 0) // every 100ms
    layer_2();
```

void l2_calc_wh(void) {
	static uint8_t ui8_1s_timmer_counter = 0;
	uint32_t ui32_temp = 0;

	if (l2_vars.ui16_battery_power_filtered_x50 > 0) {
		l2_vars.ui32_wh_sum_x5 += l2_vars.ui16_battery_power_filtered_x50 / 10;
		l2_vars.ui32_wh_sum_counter++;
	}

	// calc at 1s rate
	if (ui8_1s_timmer_counter < 10) {
		ui8_1s_timmer_counter = 0;

		// avoid zero divisison
		if (l2_vars.ui32_wh_sum_counter != 0) {
			ui32_temp = l2_vars.ui32_wh_sum_counter / 36;
			ui32_temp = (ui32_temp
					* (l2_vars.ui32_wh_sum_x5 / l2_vars.ui32_wh_sum_counter))
					/ 500;
		}

		l2_vars.ui32_wh_x10 = l2_vars.ui32_wh_x10_offset + ui32_temp;
	}
	ui8_1s_timmer_counter++;
}

**Comment by lowPerformer** [comment]
void l2_calc_wh(void) seems strange:

```
void l2_calc_wh(void) {
	static uint8_t ui8_1s_timmer_counter = 0;
	uint32_t ui32_temp = 0;

	if (l2_vars.ui16_battery_power_filtered_x50 > 0) {
		l2_vars.ui32_wh_sum_x5 += l2_vars.ui16_battery_power_filtered_x50 / 10;
		l2_vars.ui32_wh_sum_counter++;
	}

	// calc at 1s rate
	if (ui8_1s_timmer_counter < 10) {
		ui8_1s_timmer_counter = 0;

		// avoid zero divisison
		if (l2_vars.ui32_wh_sum_counter != 0) {
			ui32_temp = l2_vars.ui32_wh_sum_counter / 36;
			ui32_temp = (ui32_temp
					* (l2_vars.ui32_wh_sum_x5 / l2_vars.ui32_wh_sum_counter))
					/ 500;
		}

		l2_vars.ui32_wh_x10 = l2_vars.ui32_wh_x10_offset + ui32_temp;
	}
	ui8_1s_timmer_counter++;
}
```
Shouldn't it be
```
// calc at 1s rate
if (ui8_1s_timmer_counter >= 10) {
}
```

Maybe I can test this week. Can't promise...

**Comment by casainho** [comment]
Should be this, just like on l2_calc_odometer():

```
	// calc at 1s rate
	if (ui8_1s_timmer_counter++ >= 10) {
		ui8_1s_timmer_counter = 0;
```

```
static void l2_calc_odometer(void) {
	uint32_t uint32_temp;
	static uint8_t ui8_1s_timmer_counter;

	// calc at 1s rate
	if (ui8_1s_timmer_counter++ >= 10) {
		ui8_1s_timmer_counter = 0;
```

**Comment by lowPerformer** [comment]
Yes.
Please correct me if I am wrong (I lost myself often enough in such trivials...) but your code gets true not until the 11th iteration which means 1.1 seconds (10 % error).
Should it be
if (++ui8_1s_timmer_counter >= 10)

**Comment by casainho** [comment]
I always give that error......

**Comment by lowPerformer** [comment]
Hopefully fixed with PR #35 . Please re-open if not!
